deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -708,6 +764,8 @@ def initialize_gradient_partitioning_data_structures(self):
                 self.total_grads_in_partition[i][partition_id] = 0
                 self.initialize_gradient_partition(i, param_group, partition_id)
                 self.is_partition_reduced[i][partition_id] = False
+
+            for partition_id in range(total_partitions):
```

Comment:
Why is this needed? It seems not to change the logic.

Reply:
The next line under this (R771) calls `self.get_first_param_index`. This code in this function look up the `self.param_to_partition_ids[group_id][param_id]` (R743) in the order of param_id order, which will cause error. Because `self.param_to_partition_ids` is set in `def initialize_gradient_partitioning_data_structures`(R747) in the loop of `param_group` and `total_partitions`. For example, the first param in the first group maybe not set to first partition. So the dict `self.param_to_partition_ids[0]` will have no key of "0" when finish the first loop of `total_partition `of the first group(L704). After this change, all of  `self.param_to_partition_ids[group_id][param_id]` exists, which will avoid no key error.

Reply:
But I think `self.param_to_partition_ids[*][*]` is set by `self.initialize_gradient_partition(...)` in (R766), no? 

Or is it because you are changing the order of `self.param_to_parition_ids` initialization, so we need to wait for the entire dict to be initialized before any param_id can be safely read?

Reply:
> But I think `self.param_to_partition_ids[*][*]` is set by `self.initialize_gradient_partition(...)` in (R766), no?
Yes, exactly. It's set by `self.initialize_gradient_partition(...)`.

> Or is it because you are changing the order of `self.param_to_parition_ids` initialization, so we need to wait for the entire dict to be initialized before any param_id can be safely read?
Yes, you are right:)